### PR TITLE
Document boundary behavior of `draw.polygon` and `draw.polygon2mask`

### PR DESCRIPTION
**Pull request by lagru**
## Description

Replaces #6649.

The extended example section and other improvements hopefully make it clear how these functions behave, if polygon vertices are outside a defined coordinate range.

While playing around with this and testing the behavior, I noticed that the coordinate values returned by `polygon` always have a lower bound of 0 https://github.com/scikit-image/scikit-image/blob/23711dc749f750880c08fdfee93329e4d320231f/skimage/draw/_draw.pyx#L220-L222
How do you feel about (optionally) removing this constraint? That would make it possible to get fun "wrap-around" behavior...

I was gonna suggest removing the thin wrapper of `polygon` around `_polgyon`. That could remove quite some duplication but it seems we'd need [pytest-cython](https://pypi.org/project/pytest-cython/) for doctest support in Cython. And I don't think that testing dependency is worth it. 

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Comment by stefanv** [comment]
> How do you feel about (optionally) removing this constraint? That would make it possible to get fun "wrap-around" behavior...

The `shape` argument is usually the way we control clipping, otherwise coordinates <0 should be allowed.

**Comment by lagru** [comment]
That's also my intuition about how to solve this on the API level. If `shape` is provided, bounds are checked to be >= 0 and within the image, otherwise all bet's are off. How about this transition path

- release x: when negative polygon coordinates are detected, raise a warning that the clipping behavior will be removed in the future. Advise users to clip themselves to avoid the warning.
- release x+1: switch to new behavior.

But we should make sure that we are consistent about this with the rest of the API in `skimage.draw`. I'll move this out into its own issue to collect input there.

**Comment by lagru** [comment]
Interesting. It seem like sphinx renders the emphasized "polygon" https://github.com/scikit-image/scikit-image/blob/6164ff8565a050fdf72a9b4a8b151710b5cbd1c4/skimage/draw/_polygon2mask.py#L50

[as a link](https://output.circle-artifacts.com/output/job/a4d578f3-c724-4457-a358-ee5fc0b00fe5/artifacts/0/doc/build/html/api/skimage.draw.html#polygon2mask). That's unintended behavior, as link should refer to the parameter and not to the function.

**Comment by lagru** [comment]
I think it's because we set https://github.com/scikit-image/scikit-image/blob/23711dc749f750880c08fdfee93329e4d320231f/doc/source/conf.py#L66
([`default_role` in Sphinx's docs](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-default_role)). 

I propose that we remove that option. If we want to link an object, I think we should do so explicitly rather than implicitly if Sphinx find's a match by chance. That way the source itself also shows the intended target. E.g. it's clear that :func:\`~.polygon\` references the function and not the parameter.

File: skimage/draw/draw.py [hunk 1]
```diff
@@ -475,8 +485,27 @@ def polygon(r, c, shape=None):
            [0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
            [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
-           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
+
+    If the image `shape` is defined and vertices / points of the `polygon` are
+    outside this coordinate space, only a part (or none at all) of the polygon's
```
Comment (henrypinkard) [inline skimage/draw/draw.py:491]: If this example is meant to demonstrate the offset, then what the offset is doing should be mentioned here. I would suggest replacing with:

Defining an image `shape` can be used to draw an area of the polygon smaller or larger than the extent of the polygon's vertices. This example shows how shifting the polygon's vertices by an offset can be used to draw an arbitrary sub-region of the polygon

Comment (lagru) [reply skimage/draw/draw.py:491]: Great suggestion! I found my first sentence clearer but incorporated your second one in 6221476b7c67d677787fdc83418e8d67b2dfd23f. :)

File: skimage/draw/_draw.pyx [hunk 2]
```diff
@@ -193,25 +193,32 @@ def _line_aa(Py_ssize_t r0, Py_ssize_t c0, Py_ssize_t r1, Py_ssize_t c1):
...
     -------
     rr, cc : ndarray of int
         Pixel coordinates of polygon.
         May be used to directly index into an array, e.g.
         ``img[rr, cc] = 1``.
+
+    Notes
+    -----
+    This function ensures that `rr` and `cc` don't contain negative values.
+    Pixels of the polygon that whose coordinates are smaller 0, are not drawn.
```
Comment (stefanv) [inline skimage/draw/_draw.pyx:221]:
```suggestion
    Pixels in the polygon with coordinates smaller than 0 are not drawn.
```

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
Thank you!

**Pull request merged**